feat: expose insights into user activity

mtojek opened the pull request:
Related: https://github.com/coder/coder/issues/9497

This PR extends Insights API with the user activity. BTW I moved golden files for `TestTemplateInsights_Golden` to a dedicated subdirectory. 

_Note: I know that this PR is large, mainly due to generated golden files. I'm happy to split the PR into pieces if it helps!_
 

TODO:
- [x] model structures in codersdk
- [x] coderd logic
- [x] database query
- [x] dbfake logic
- [x] unit tests

mtojek committed:
codersdk

mtojek committed:
stub for user-activity

mtojek committed:
WIP

mtojek committed:
fixes

mtojek committed:
dbauthz

mtojek committed:
Adjust SQL query

mtojek commented on coderd/database/queries/insights.sql:
```diff
@@ -21,6 +21,86 @@ WHERE
 GROUP BY workspace_agent_stats.user_id, users.username, users.avatar_url
 ORDER BY user_id ASC;
 
+-- name: GetUserActivityInsights :many
```
@mafredri Could you please take a look and validate if the query makes sense? If so, I will implement the matching`dbfake` part.

mafredri replied:
Had one comment but otherwise this looks mostly correct to me! Only had the one comment.

mtojek committed:
Fixes

mafredri commented on coderd/database/queries/insights.sql:
```diff
@@ -21,6 +21,86 @@ WHERE
+		template_id,
+		start_time,
+		seconds
+	FROM combined_stats
```
You could remove this additional step if you change `UNION ALL` to `UNION` above, it'll remove duplicates.

However, in this instance I'm not sure if we should sum template IDs separately. Say you use template A and B concurrently for 1h, you will be listed as 2h in the stats now. Perhaps instead we should modify this distinct to be a group by (user, start time) with array_agg on the templates.

mtojek committed:
Fix: combined

mtojek committed:
Implement dbfake

mtojek committed:
Merge branch 'main' into 9497-db-first

mtojek committed:
RBAC tests

mtojek committed:
Refactor tests

mtojek committed:
More tests

mtojek committed:
TestUserActivityInsights

mtojek committed:
fix: empty line

mtojek committed:
fix

mtojek committed:
Merge branch 'main' into 9497-db-first

mtojek committed:
Golden files

mtojek committed:
lint

mtojek committed:
linter

mtojek committed:
Fix user IDs

mtojek committed:
cleanup

mtojek commented:
@mafredri Let's start the next review round.

I implemented the entire workflow and covered responses with golden files. My biggest concern is `TestUserActivityInsights_Golden` function which is a fork/ripoff of the other large function. Do you think that I should extract common structures/functions? I added a few changes:
* disabled/removed parameters
* use `dbgen.*` to create users/membership/tokens

mtojek committed:
Merge branch 'main' into 9497-db-first

mafredri commented on coderd/testdata/insights/user-activity/multiple_users_and_workspaces_week_all_templates.json.golden:
```diff
@@ -0,0 +1,41 @@
+        "user_id": "00000000-0000-0000-0000-000000000004",
+        "username": "user1",
+        "avatar_url": "",
+        "seconds": 30540
```
Note-to-self: Also verified this number, 1h is lost from `users[0].workspaces[1]` due to timeframe and all the minute intervals are summed from `users[0].workspaces[0]` giving us 8h + 29m. The hours in u0w0 overlap with u0w1 so we don't see an increase from those.

mafredri reviewed (approved):
Looks good! Don't worry about refactoring to remove duplication for now. We can defer it until there's 1-2 more use-cases or when we start having trouble keeping the data in sync.

mtojek committed:
Try: truncate 1 min

mtojek committed:
Note

mtojek commented on coderd/database/queries/insights.sql:
```diff
@@ -21,6 +21,79 @@ WHERE
+-- name: GetUserActivityInsights :many
+-- GetUserActivityInsights returns the ranking with top active users.
+-- The result can be filtered on template_ids, meaning only user data from workspaces
+-- based on those templates will be included.
```
Done. I added a comment close to the SQL function.

mtojek commented on coderd/database/dbfake/dbfake.go:
```diff
@@ -2831,6 +2831,142 @@ func (q *FakeQuerier) GetUnexpiredLicenses(_ context.Context) ([]database.Licens
+		t := s.SessionStartedAt.Truncate(5 * time.Minute)
+		if t.Before(arg.StartTime) {
+			t = arg.StartTime
+		}
```
Fixed

mtojek committed:
Agent
